### ISSUE TITLE
Sort out handling of constant features and ascertainment correction

### ISSUE THREAD

**Issue by lmaurits** [lmaurits/BEASTling#142]
Currently, BEASTling removes constant features by default, and does not do ascertainment correction for this.  We _do_ do ascertainment correction for binary recoding.  We need to:

1. Revisit the decision to remove constant features by default
2. Regardless of decision on the above, come up with a complete and consistent approach to ascertainment correction which automatically does the correct thing for each situation as far as this is possible, and makes it straightforward for the user to manually do this when automating it is not possible.

This may involve changing the semantics of one or both of the "ascertained" and "binarised" options on BinaryModel.

**Comment by lmaurits** [comment]
In order to know what to do re: ascertainment, BEASTling needs to know up to two things:

1. Have constant features been excluded from the dataset a priori? (always required)
2. If the data is binary, is this "real binary data" or the result of recoding a multistate feature as binary features (irrelevant for multistate features)

BEASTling can know 1. beyond all doubt if the data includes constant features.  If the data does _not_, there needs to be a way for the user to convey whether this is true or not, and also BEASTling needs a sensible default assumption if the user does not provide this.  This is probably going to require a new option on BaseModel, although possibly the current `ascertained` option on BinaryModel could be repurposed for this.  As for the default assumption...I would lean toward yes, as in general thing don't end up a feature in WALS etc. unless they vary.

BEASTling can know 2. beyond all doubt if it is responsible for the recoding (i.e. if the user provides multistate data for BinaryCovarion).  If it is not responsible, then we need a way for the user to convey whether this is true or not.  This is _not_ the current use of the `binarised` option in BinaryModel, but I'm starting to think that perhaps it should be?   The current use is to bypass BEASTling inspecting whether or not there are more than 2 values for any feature, which is not really very useful)?   There's also the issue of a default assumption if the user does not provide a value - and, indeed, the issue of whether we _want_ a default in this case or whether we want the `binarised` argument to be obligatory.

**Comment by lmaurits** [comment]
After a whole day of thinking about this and experimenting with various options, I've had an epiphany and this is actually embarrassingly straightforward to handle.

We absolutely should retain the current behaviour of automatically removing constant features because, in general, *we have no way to model them*.  For Mk and BSVS, the space of possible values is derived from the dataset.  If a feature is constant and you ask it to be modelled with Mk, BEASTling and BEAST will comply, but the resulting model is useless - it has likelihood 1.0 regardless of mutation rates, clock rates, branch lengths, anything.  So, the estimated mutation rate is about 1.0 (the compulsory mean), meaning leaving it in does _not_ have the naively desired effect of better informing the model about rate variation.  If you ask a constant feature to be modelled with BSVS, BEAST will complain because it tries to build an array of n*(n-1) = 0 relative rate variables.  Retaining constant features _only_ makes sense for models which define their own statespace, independent of the data.  At the moment, the only such model for BEASTling is BinaryCovarion (even if a particular feature is, say, all 1s in the data file, the model knows that there is this alternative thing 0 out there, so the model is actually meaningful).

In light of this, I think the `remove_constant_features` option should be _removed_ from BaseModel and moved to BinaryModel, as it makes no sense where it currently is.  This has simplifying flow on consequences: for Mk and BSVS, we don't need a way for the user to convey whether or not to do ascertainment correction for non-constant data, it _always_ needs to be done if timing is an issue.

BinaryModel is really the only place any of this matters, then.  A constant feature _does_ make sense there, and it _can_ inform the distribution over rates.  However, based on experimentation today, I would argue to still leave them out by default.  This is because I believe the CovarionModel (the only current subclass of BinaryModel) is too complicated to make this approach reliable: it can explain a constant feature either by giving it a very low mutation rate (the intended effect) or, if there are enough constant features, by increasing the Covarion switching rate and making the "slow" state _very_ slow compared to the "fast" state (an unintended alternative effect which may result in a constant feature receiving a not-particularly-slow mutation rate).

If we go with this approach, the only time constant features can end up in any model is if the user explicitly sets `remove_constant_features=False`.  And it seems sensible to assume that they will only do this if they actually have some constant features in their data.  In which case we definitely don't want to do ascertainment correction for non-constant features.  So, the existing `remove_constant_features` option can play the role of controlling whether or not to do that correction.

I think this only leaves one open question, which is whether or not we are happy to repurpose `binarised' to indicate whether or not data supplied to BinaryModel which already has only two values is the result of the user having already binarised multistate data outside of BEASTling.

**Comment by xrotwang** [comment]
Your solution - including repurposing `binarised` - sounds sensible to me. I also think this is something to keep in mind for efforts like CLDF: There should be standard ways to encode (and access) state space (or domain) for any parameter in a dataset. It also highlights a good use case for something like a Gramaticon (the analogy of [Concepticon](http://concepticon.clld.org) for structural features): A Gramaticon could be the place where
- information about the domain of features is curated and can be obtained,
- information about things like sensible binarizations of multi-state data can be curated or at least attached to UUIDs of features.

**Comment by lmaurits** [comment]
(ignore accidental closure)

I have done a push which I think implements everything discussed, and updated the documentation appropriately.  I want to write some more tests to make sure everything is happening as expected before I close this, however.

**Comment by lmaurits** [comment]
More than half a year later, I don't remember what those tests I wanted to write were.  But looking back through the history, I notice that one day after making the above comment I committed several tests related to exactly this issue, so I'm going to assume that I did in fact take care of this and just forgot to close the issue.  I'll do so now, in the expectation that the process of closing #161 will shake out any forgotten problems.
